File: arc011/prob3.py
import sys
from collections import deque

def sibling(w1, w2):
    c = 0
    for c1, c2 in zip(w1, w2):
        if c1 != c2:
            c += 1
            if c > 1:
                return False
    return c == 1

def answer(first, last, route):
    print len(route)
    print first
    for w in route:
        print w
    print last
    sys.exit()

def main():
    readline = sys.stdin.readline
    first, last = readline().split()
    N = int(readline())
    words = [readline().strip() for _ in range(N)]

    if first == last:
        answer(first, last, ())

    unused = set(words)
    unused.add(last)
    que = deque()
    que.append((first, ()))

    while que:
        word, route = que.popleft()
        for ww in list(unused):
            if sibling(word, ww):
                unused.remove(ww)
                if ww == last:
                    answer(first, last, route)
                else:
                    que.append((ww, route + (ww,)))
    else:
        print -1

main()

File: arc011/prob3.cxx
#include <iostream>
#include <set>
#include <map>
#include <utility>
#include <vector>
#include <queue>

using namespace std;

bool sibling(const string& w1, const string& w2) {
    int c = 0;
    for (int i = 0; i < w1.size(); ++i) {
        if (w1[i] != w2[i]) {
            c++;
            if (c > 1) return false;
        }
    }
    return c == 1;
}

int main()
{
    int N;
    string first, last;
    set<string> unused;

    cin >> first >> last >> N;

    for (int i = 0; i < N; ++i) {
        string w;
        cin >> w;
        unused.insert(w);
    }

    if (first == last) {
        cout << 0 << '\n';
        cout << first << '\n' << last << '\n';
        return 0;
    }

    queue<pair<string, vector<string> > > que;
    que.push(make_pair(first, vector<string>()));

    while (!que.empty()) {
        auto v = que.front();
        que.pop();

        string &word = v.first;

        if (sibling(word, last)) {
            cout << v.second.size() << "\n" << first << "\n";
            for (int i=0; i<v.second.size(); ++i) {
                cout << v.second[i] << "\n";
            }
            cout << last << "\n";
            return 0;
        }

        set<string>::iterator it = unused.begin(), ite = unused.end();
        while (it != ite) {
            if (sibling(word, *it)) {
                vector<string> route = v.second;
                route.push_back(*it);
                que.push(make_pair(*it, route));
                unused.erase(it++);
            } else {
                ++it;
            }
        }
    }

    cout << "-1\n";
}
